src/txn/txn_timestamp.c

```diff
@@ -845,140 +842,182 @@ __wt_txn_parse_prepare_timestamp(
-		ts_oldest = txn_global->oldest_timestamp;
-		if (ts < ts_oldest) {
-			/*
-			 * If given read timestamp is earlier than oldest
-			 * timestamp then round the read timestamp to
-			 * oldest timestamp.
-			 */
-			if (round_to_oldest)
-				txn->read_timestamp = ts_oldest;
-			else {
-				__wt_readunlock(session, &txn_global->rwlock);
-				__wt_timestamp_to_string(ts, ts_string[0]);
-				__wt_timestamp_to_string(
-				    ts_oldest, ts_string[1]);
-				WT_RET_MSG(session, EINVAL, "read timestamp "
-				    "%s older than oldest timestamp %s",
-				    ts_string[0], ts_string[1]);
-			}
-		} else {
-			txn->read_timestamp = ts;
-			/*
-			 * Reset to avoid a verbose message as read
-			 * timestamp is not rounded to oldest timestamp.
-			 */
-			round_to_oldest = false;
-		}
+		__wt_timestamp_to_string(read_ts, ts_string[0]);
+		__wt_timestamp_to_string(ts_oldest, ts_string[1]);
+		__wt_verbose(session, WT_VERB_TIMESTAMP, "read "
+		    "timestamp %s : rounded to oldest timestamp %s",
+		    ts_string[0], ts_string[1]);
+	}
 
-		__wt_txn_set_read_timestamp(session);
-		__wt_readunlock(session, &txn_global->rwlock);
-		if (round_to_oldest &&
-		    WT_VERBOSE_ISSET(session, WT_VERB_TIMESTAMP)) {
-			/*
-			 * This message is generated here to reduce the span of
-			 * critical section.
-			 */
-			__wt_timestamp_to_string(ts, ts_string[0]);
-			__wt_timestamp_to_string(ts_oldest, ts_string[1]);
-			__wt_verbose(session, WT_VERB_TIMESTAMP, "Read "
-			    "timestamp %s : Rounded to oldest timestamp %s",
-			    ts_string[0], ts_string[1]);
-		}
+	/*
+	 * If we already have a snapshot, it may be too early to match
+	 * the timestamp (including the one we just read, if rounding
+	 * to oldest).  Get a new one.
+	 */
+	if (F_ISSET(txn, WT_TXN_RUNNING))
+		__wt_txn_get_snapshot(session);
 
-		/*
-		 * If we already have a snapshot, it may be too early to match
-		 * the timestamp (including the one we just read, if rounding
-		 * to oldest).  Get a new one.
-		 */
-		if (F_ISSET(txn, WT_TXN_RUNNING))
-			__wt_txn_get_snapshot(session);
+	return (0);
+}
+
+/*
+ * __wt_txn_set_timestamp --
+ *	Parse a request to set a timestamp in a transaction.
+ */
+int
+__wt_txn_set_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
+{
+	WT_CONFIG_ITEM cval;
+	WT_DECL_RET;
+	WT_TXN *txn = &session->txn;
+	wt_timestamp_t ts;
+
+	WT_TRET(__wt_txn_context_check(session, true));
+
+	/* Look for round_to_oldest configuration. */
+	ret = __wt_config_gets_def(session, cfg, "round_to_oldest", 0, &cval);
+	if (cval.val)
+		F_SET(txn, WT_TXN_TS_ROUND_READ);
+
+	/* Look for a commit timestamp. */
+	ret = __wt_config_gets_def(session, cfg, "commit_timestamp", 0, &cval);
+	WT_RET_NOTFOUND_OK(ret);
+	if (ret == 0 && cval.len != 0) {
+		WT_RET(__wt_txn_parse_timestamp(session, "commit", &ts, &cval));
+		WT_RET(__wt_txn_set_commit_timestamp(session, ts));
+		__wt_txn_publish_commit_timestamp(session);
+	}
+
+	/* Look for a durable timestamp. */
```

Comment:
I'd like this comment to mention that it's important for it to come after setting the commit timestamp.